runners/core-java/src/main/java/org/apache/beam/runners/core/WatermarkHold.java

```diff
@@ -247,8 +249,26 @@ private Instant addGarbageCollectionHold(
   /**
    * Prefetch watermark holds in preparation for merging.
    */
-  public void prefetchOnMerge(MergingStateAccessor<?, W> state) {
-    StateMerging.prefetchWatermarks(state, elementHoldTag);
+  public void prefetchOnMerge(MergingStateAccessor<?, W> context) {
```

Comment:
Is there any specific reason why code is moved out from StateMerging.java?

Reply:
I moved the merging code out of StateMerging.java so that I could reuse WatermarkHold.addElementHold instead of passing in the watermarks to StateMerging and duplicating the late detection logic.  To keep everything located together I moved the other uses of WatermarkHoldState from StateMerging to WatermarkHold.  AFAIK the StateMerging class is a leftover from lazy state merging and doesn't provide much benefit any more.

Reply:
As I am not familiar with the StateMerging, I will leave it up to the committer who will merge this PR in.